pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh

```diff
@@ -507,11 +520,97 @@ if [ "$action" = dry-build ]; then
     extraBuildFlags+=(--dry-run)
 fi
 
+if [ "$action" = list-generations ]; then
+    if [ ! -L "$profile" ]; then
+        log "No profile \`$(basename "$profile")' found"
+        exit 1
+    fi
+
+    generation_from_dir() {
+        generation_dir="$1"
+        generation_base="$(basename "$generation_dir")" # Has the format "system-123-link" for generation 123
+        no_link_gen="${generation_base%-link}"  # remove the "-link"
+        echo "${no_link_gen##*-}" # remove everything before the last dash
+    }
+    describe_generation(){
+        generation_dir="$1"
+        generation_number="$(generation_from_dir "$generation_dir")"
+        nixos_version="$(cat "$generation_dir/nixos-version" 2> /dev/null || echo "Unknown")"
+
+        kernel_dir="$(dirname "$(realpath "$generation_dir/kernel")")"
+        kernel_version="$(ls "$kernel_dir/lib/modules" || echo "Unknown")"
+
+        configurationRevision="$("$generation_dir/sw/bin/nixos-version" --configuration-revision 2> /dev/null || true)"
+
+        # Old nixos-version output ignored unknown flags and just printed the version
+        # therefore the following workaround is done not to show the default output
+        nixos_version_default="$("$generation_dir/sw/bin/nixos-version")"
+        if [ "$configurationRevision" == "$nixos_version_default" ]; then
+             configurationRevision=""
+        fi
+
+        if [ -z "$json" ]; then
+            build_date="$(date --date="@$(stat "$generation_dir" --format=%W)" "+%a %F %T")"
+        else
+            # jq automatically quotes the output => don't try to quote it in output!
+            build_date="$(stat "$generation_dir" --format=%W | jq 'todate')"
+        fi
+
+        if [ -z "$json" ]; then
+            unset current_generation_tag
+        else
+            current_generation_tag="false"
+        fi
+        if [ "$(basename "$generation_dir")" = "$(readlink "$profile")" ]; then
+            if [ -z "$json" ]; then
+                current_generation_tag="  (current)"
+            else
+                current_generation_tag="true"
+            fi
+        fi
+
+        if [ -z $json ]; then
+            echo "$generation_number,\
+$build_date,\
+$nixos_version,\
+$kernel_version,\
+$configurationRevision\
+$current_generation_tag"
+        else
+            # Escape userdefined strings
+            nixos_version="$(jq -aR <<< "$nixos_version")"
+            kernel_version="$(jq -aR <<< "$kernel_version")"
+            configurationRevision="$(jq -aR <<< "$configurationRevision")"
+            echo "{ \
+  \"generation\": $generation_number, \
+  \"date\": $build_date, \
+  \"nixosVersion\": $nixos_version, \
+  \"kernelVersion\": $kernel_version, \
+  \"configurationRevision\": $configurationRevision, \
+  \"current\": $current_generation_tag\
+}"
+        fi
+    }
+
+    find "$(dirname "$profile")" -regex "$profile-[0-9]+-link" |
+        sort -rg |
```

Comment:
FIXME: I thought the `-g` would take care of 1 < 2 < 10, but it is just ordered by first digit.

Reply:
```
Generation  Build-date               NixOS version           Kernel  Configuration Revision
9           Tue 2023-06-20 15:03:54  23.11.20230618.09c81a9  6.3.8   dirty
8           Tue 2023-06-20 14:27:04  23.11.20230618.09c81a9  6.3.8   40004285a151ec03048c826bc82b5fa34cf73ff0
7           Tue 2023-06-20 14:14:45  23.11.20230618.09c81a9  6.1.34  dirty
6           Tue 2023-06-20 14:05:13  23.11.20230618.09c81a9  6.1.34  dirty
5           Tue 2023-06-20 13:16:19  23.11.20230618.09c81a9  6.3.8   dirty
4           Tue 2023-06-20 12:58:30  23.11.20230618.09c81a9  6.3.8   dirty
3           Tue 2023-06-20 11:53:36  23.05.1139.572d2693045  6.1.34
2           Tue 2023-06-20 10:03:19  23.05.1139.572d2693045  6.1.34
1           Mon 2023-06-19 16:05:32  23.05.1123.aaef163eac7  6.1.34
11          Tue 2023-06-20 18:52:13  23.11.20230618.09c81a9  6.3.8   0f020d3fbaae0e0972911a0a786891998c58ef03  (current)
10          Tue 2023-06-20 18:43:05  23.11.20230618.09c81a9  6.3.8   e6dc04fa457e207b19c62bb3b66dd2b1796a9417
```